test/integration/scheduler_perf/dra.go

```diff
@@ -363,3 +373,307 @@ claims:
+	DurationParam string
+	// ScheduledParam is the parameter for the expected scheduled state.
+	// If set, the expected scheduled state is parameterized through this parameter.
+	ScheduledParam string
+	// ExpectedScheduled is the expected scheduled state.
+	// If true, the op checks if all pods are scheduled.
+	// If false, the op checks if all pods are unscheduled.
+	// This is used to test the device binding conditions.
+	ExpectedScheduled bool
+}
+
+func (op *checkPodScheduledOp) isValid(allowParameterization bool) error {
+	if op.Namespace == "" {
+		return fmt.Errorf("namespace must be set")
+	}
+	return nil
+}
+
+func (op *checkPodScheduledOp) collectsMetrics() bool {
+	return false
+}
+
+func (op *checkPodScheduledOp) patchParams(w *workload) (realOp, error) {
+	if op.ScheduledParam != "" {
+		var err error
+		op.ExpectedScheduled, err = getParam[bool](w.Params, op.ScheduledParam[1:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if op.DurationParam != "" {
+		durationStr, err := getParam[string](w.Params, op.DurationParam[1:])
+		if err != nil {
+			return nil, err
+		}
+		if op.Duration.Duration, err = time.ParseDuration(durationStr); err != nil {
+			return nil, fmt.Errorf("parsing duration parameter %s: %w", op.DurationParam, err)
+		}
+	}
+
+	return op, op.isValid(false)
+}
+
+func (op *checkPodScheduledOp) requiredNamespaces() []string {
+	if op.Namespace != "" {
+		return []string{op.Namespace}
+	}
+	return nil
+}
+
+func (op *checkPodScheduledOp) run(tCtx ktesting.TContext) {
+	ctx, cancel := context.WithTimeout(tCtx, op.Duration.Duration)
+	defer cancel()
+
+	var (
+		lastViolation string
+		stableStart   *time.Time
+	)
+
+	validate := func(pod v1.Pod) bool {
+		isScheduled := pod.Spec.NodeName != ""
+		return (op.ExpectedScheduled && isScheduled) || (!op.ExpectedScheduled && !isScheduled)
+	}
+
+	err := wait.PollUntilContextTimeout(ctx, 1*time.Second, op.Duration.Duration, true, func(ctx context.Context) (bool, error) {
+		pods, err := tCtx.Client().CoreV1().Pods(op.Namespace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			tCtx.Logf("Failed to list pods: %v", err)
+			return false, nil
+		}
+
+		allValid := true
+		currentViolation := ""
+		for _, pod := range pods.Items {
+			if !validate(pod) {
+				allValid = false
+				currentViolation = fmt.Sprintf("Pod %s: ExpectedScheduled=%v, ActualScheduled=%v (Node=%s)",
+					pod.Name, op.ExpectedScheduled, pod.Spec.NodeName != "", pod.Spec.NodeName)
+				break
+			}
+		}
+
+		if !allValid {
+			lastViolation = currentViolation
+			stableStart = nil
+			return false, nil
+		}
+
+		if op.ExpectedScheduled {
+			tCtx.Logf("All pods scheduled successfully")
+			return true, nil
+		} else {
+			if stableStart == nil {
+				now := time.Now()
+				stableStart = &now
+				tCtx.Logf("All pods are unscheduled, starting stability check")
+				return false, nil
+			}
+
+			if time.Since(*stableStart) >= 5*time.Second {
```

Comment:
This doesn't get reached if op.Duration.Duration is shorter than 5 seconds. Instead it times out and fails below with "Pods were unexpectedly scheduled. Last violation: ".

Can probably be fixed with a check whether `lastViolation` is empty.

Reply:
I add the empty check for `lastViolation`.